### DISCUSSION TITLE
expression: required field missing; paths: unknown field)

### DISCUSSION THREAD

**Discussion by tirelibirefe** [Kong/kong#12044] [Help] [answered]
Hello,
Based on the samples [here](https://docs.konghq.com/gateway/latest/get-started/services-and-routes/) ; I defined a service by using admin api and I would like to define a route for it but I get following error:

```
curl -i -X POST $APIURL/services/httpbin/routes \
  --data name=httpbin \
  --data 'paths[]=/httpbin'
HTTP/2 400 
date: Thu, 16 Nov 2023 07:55:42 GMT
content-type: application/json; charset=utf-8
content-length: 200
access-control-allow-origin: *
access-control-allow-credentials: true
x-kong-admin-latency: 10

{"name":"schema violation","message":"2 schema violations (expression: required field missing; paths: unknown field)","code":2,"fields":{"expression":"required field missing","paths":"unknown field"}}
```

When I try to define a route by using Postman, I get following error:

```
{
    "name": "schema violation",
    "message": "10 schema violations (client_certificate: unknown field; connect_timeout: unknown field; expression: required field missing; host: unknown field; path: unknown field; port: unknown field; protocol: unknown field; read_timeout: unknown field; retries: unknown field; write_timeout: unknown field)",
    "code": 2,
    "fields": {
        "retries": "unknown field",
        "connect_timeout": "unknown field",
        "read_timeout": "unknown field",
        "port": "unknown field",
        "write_timeout": "unknown field",
        "client_certificate": "unknown field",
        "expression": "required field missing",
        "host": "unknown field",
        "path": "unknown field",
        "protocol": "unknown field"
    }
}
```

If I try define a route by using Manager, I get the following error:

![image](https://github.com/Kong/kong/assets/45227339/675c6d9d-d4b7-4026-af75-ede77ea3a871)

Could you please advise the correct way defining a route or what I am missing?

Thanks & Regards

**Comment by chronolaw** [comment, marked as answer]
It looks like you are using kong gateway with router_flavor `expressions`. 
With this flavor we use the field `expression` instead of `paths` `methods` and so on.

Please refer to https://docs.konghq.com/gateway/3.5.x/key-concepts/routes/expressions/.

**Comment by chronolaw** [comment]
You can check the current flavor by this command:
```
curl 127.1:8001 | jq '.configuration.router_flavor'
```

**Comment by tirelibirefe** [comment]
Hello @chronolaw 
Thanks for the reply and advises.
Does it mean that if I use expressions as router flavour, I can't use Kong Manager UI to define new routes?
I will not use Kong as an ingress controller. Is there any advantages to use expressions instead of traditional router? 

Thanks & Regards

**Reply by chronolaw** [reply]
Kong Manager supports expression, I think that we can write expressions in UI.

The expression router is faster and more powerful than traditional router, it is built by Rust, and supports http queries as match condition.

**Reply by zsh1313** [reply]
hello @chronolaw 
I had same problem as above, and found your reply here, really helpful. Thank you

I have the router_flavor configured to "expressions". But in kong manager GUI, there is no "expression" field to write expressions. 

Is there anything that needs to be configured to enable expressions in kong manager GUI? thank you

**Reply by chronolaw** [reply]
@sumimakito, could you answer this kong manager GUI's question? thanks.

**Reply by ligreman** [reply]
I can confirm that there is no expressions management in current Kong Manager. There is no field to edit it. But in the Route details the expression field is shown.

So at least two fields are needed:
-  "expression" field
- "expression priority" field, to define the priority value of the expression (integer)

Thanks.

PD: a guided or wizard group of fields would be optimal, to improve usability. In my own interface I made this form to manage the expressions more visually (you can build the expression with the fields and buttons, without really typing it). It's made for non developer administrators, of course:

![a](https://github.com/Kong/kong/assets/2909815/d267ca76-69d7-4bc7-bfb3-577f478801dc)
![b](https://github.com/Kong/kong/assets/2909815/1c47c93b-63f7-42e8-9908-a5a2e790b183)

**Comment by sumimakito** [comment]
Hello everyone in this discussion, since https://github.com/Kong/kong-manager/pull/217, we have added an interactive editor to create/edit routes defined in the Expressions language to Kong Manager for Kong Gateway CE. 

With `KONG_ROUTER_FLAVOR` (env) or `router_flavor` (kong.conf) set to `expressions`, you can switch to the "Expressions" tab to try this out:

<img width="1741" alt="Screenshot 2024-08-21 at 13 21 18" src="https://github.com/user-attachments/assets/c9f485df-a052-4e71-a44e-4129304475b0">

Please let us know if you have further questions.
